src/uct/ib/ud/verbs/ud_verbs.c

```diff
@@ -393,6 +393,20 @@ uct_ud_verbs_iface_poll_tx(uct_ud_verbs_iface_t *iface, int is_async)
     return 1;
 }
 
+static UCS_F_ALWAYS_INLINE size_t uct_ud_iface_get_gid_len(void *packet)
+{
+  /* The GRH will contain either an IPv4 or IPv6 header. If the former is
+   * present the header will start at offset 20 in the buffer otherwise it
+   * will start at offset 0. Since the two headers are of fixed size (20 or
+   * 40 bytes) this means we will either see 0x6a at offset 0 (IPv6) or 0x45
+   * at offset 20. The detection is a little tricky for IPv6 given that the
+   * first 20B are undefined for IPv4. To overcome this the first byte of
+   * the posted receive buffer is set to 0xff.
+   */
+    return (((unsigned char *)packet)[0] == 0x6a) ? UCS_IPV6_ADDR_LEN :
```

Comment:
it would be safer to check just version in the first half-byte of the ipv6 header

Reply:
True, but two things. 1) IPv6 headers are a minimum of 40 bytes and 2) the GRH is always 40 bytes. I can make the change but the more explicit the check the less likely it will match a random byte if the provider does something unexpected.

Reply:
but in ipv6 the first byte contains a part of traffic class, which can potentially vary (it's not a header len like in ipv4).

Reply:
Ah, you are correct. I swear I was looking at the right document but it is indeed traffic class. Will fix.